File: num_bad_pairs.rs
// 2364. Count Number of Bad Pairs
use std::collections::HashMap;
struct Solution;

/*
 * you are given a 0-indexed integer array 'nums'. a pair of indices '(i, j)' is
 * a bad pair if 'i < j' and 'j - i != nums[j] - nums[i]'. return the total
 * number of bad pairs in 'nums'.
 */

impl Solution {
    pub fn count_bad_pairs(nums: Vec<i32>) -> i64 {
        let mut hm: HashMap<i32, i32> = HashMap::with_capacity(nums.len());
        let mut ans = 0;
        nums.iter().zip(0..).for_each(|(&x, i)| {
            let delta = x - i;
            let e = hm.entry(delta).or_default();
            ans += (i - *e) as i64;
            *e += 1;
        });
        ans
    }
}

fn main() {
    let n1 = vec![4, 1, 3, 3];
    let n2 = vec![1, 2, 3, 4, 5];
    println!("{}", Solution::count_bad_pairs(n1)); //expect: 5
    println!("{}", Solution::count_bad_pairs(n2)); //expect: 0
}

File: num_bad_pairs.py
# 2364. Count Number of Bad Pairs
from collections import defaultdict

"""
you are given a 0-indexed integer array 'nums'. a pair of indices '(i, j)' is
a bad pair if 'i < j' and 'j - i != nums[j] - nums[i]'. return the total
number of bad pairs in 'nums'.
"""


class Solution(object):
    def countBadPairs(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        mp, ans = defaultdict(int), 0
        for i in range(len(nums)):
            ans += i - mp[nums[i] - i]
            mp[nums[i] - i] += 1
        return ans


if __name__ == "__main__":
    obj = Solution()
    print(obj.countBadPairs(nums=[4, 1, 3, 3]))
    print(obj.countBadPairs(nums=[1, 2, 3, 4, 5]))
